x-pack/plugins/licensing/server/plugin.ts

```diff
@@ -4,145 +4,142 @@
-      .create<LicensingConfigType | { config: LicensingConfigType }>()
-      .pipe(
-        map(config =>
-          'config' in config
-            ? new LicensingConfig(config.config, this.context.env)
-            : new LicensingConfig(config, this.context.env)
-        )
-      );
+    this.config$ = this.context.config.create<LicenseConfigType>();
   }
 
-  private hasLicenseInfoChanged(newLicense: any) {
-    const currentLicense = this.poller.subject$.getValue();
+  public async setup(core: CoreSetup) {
+    this.logger.debug('Setting up Licensing plugin');
+    const config = await this.config$.pipe(take(1)).toPromise();
+    const dataClient = await core.elasticsearch.dataClient$.pipe(take(1)).toPromise();
 
-    if ((currentLicense && !newLicense) || (newLicense && !currentLicense)) {
-      return true;
-    }
+    const { refresh, license$ } = this.createLicensePoller(dataClient, config.pollingFrequency);
 
-    return (
-      newLicense.type !== currentLicense.type ||
-      newLicense.status !== currentLicense.status ||
-      newLicense.expiry_date_in_millis !== currentLicense.expiryDateInMillis
-    );
+    core.http.registerRouteHandlerContext('licensing', createRouteHandlerContext(license$));
+
+    return {
+      refresh,
+      license$,
+    };
   }
 
-  private async fetchInfo(core: CoreSetup, clusterSource: string, pollingFrequency: number) {
-    this.logger.debug(
-      `Calling [${clusterSource}] Elasticsearch _xpack API. Polling frequency: ${pollingFrequency}`
+  private createLicensePoller(clusterClient: IClusterClient, pollingFrequency: number) {
+    const manualRefresh$ = new Subject();
+    const intervalRefresh$ = timer(0, pollingFrequency);
+    const refresh$ = merge(intervalRefresh$, manualRefresh$).pipe(takeUntil(this.stop$));
+
+    const { update$, fetchSubscription } = createLicenseUpdate(refresh$, () =>
+      this.fetchLicense(clusterClient)
     );
 
-    const cluster = await core.elasticsearch.dataClient$.pipe(first()).toPromise();
+    this.licenseFetchSubscription = fetchSubscription;
+
+    const license$ = update$.pipe(
+      tap(license => {
+        this.logger.debug(
+          'Imported license information from Elasticsearch:' +
+            [
+              `type: ${license.type}`,
+              `status: ${license.status}`,
+              `expiry date: ${moment(license.expiryDateInMillis, 'x').format()}`,
+            ].join(' | ')
+        );
+      })
+    );
+
+    return {
+      refresh: () => {
+        this.logger.debug('Requesting Elasticsearch licensing API');
+        manualRefresh$.next();
+      },
+      license$,
+    };
+  }
 
+  private fetchLicense = async (clusterClient: IClusterClient): Promise<ILicense> => {
     try {
-      const response = await cluster.callAsInternalUser('transport.request', {
+      const response = await clusterClient.callAsInternalUser('transport.request', {
         method: 'GET',
         path: '/_xpack',
```

Comment:
we need to switch `_license` endpoint. `/_xpack` will be removed in the next major release https://github.com/elastic/elasticsearch/issues/35958